port/zos390/omrtime.c

```diff
@@ -63,16 +78,22 @@ omrtime_msec_clock(struct OMRPortLibrary *portLibrary)
 uintptr_t
 omrtime_usec_clock(struct OMRPortLibrary *portLibrary)
 {
-	int64_t microsec = MAXPREC() / 8;
+	int64_t microsec = MAXPREC() / OMRPORT_TIME_HIRES_MICROTIME_DIVISOR;
 	return microsec;
 }
 
 uint64_t
 omrtime_current_time_nanos(struct OMRPortLibrary *portLibrary, uintptr_t *success)
 {
 	*success = 1;
-	uint64_t nanos = MAXPREC() * 125;
-	return nanos;
+
+	const int64_t subnanosec = MAXPREC();
+
+	// calculate nanosec = subnanosec * NUMERATOR / DENOMINATOR via integer arithmetics
+	const int64_t nanosec = muldiv64(subnanosec,
```

Comment:
Charlie, it *is* more expensive but it is also functionally correct (unlike the previous version that didn't have enough precision).

First, `OMRPORT_TIME_HIRES_NANOTIME_NUMERATOR / OMRPORT_TIME_HIRES_NANOTIME_DENOMINATOR` does not become a compile-time constant (at least, it didn't in my experiments). Even if it was, it would evaluate to `125 / 256 = 0`.

So, the order of the operations is 1) `MAXPREC() * NANOTIME_NUMERATOR` and then 2) `res / NANOTIME_DENUMERATOR`; there is a possibility of overflow in 1).